### PR TITLE
Allow custom pager support in patronictl edit-config 

### LINKED ISSUE #1695: patronictl edit-config fails under Windows
The python library `cdiff` (or as it's now known, `ydiff`) is hardwired to use `less` in version <1.2.0
Regular Windows installs only have `more` and that is not enough.

When running `edit-config` in quiet mode, no diffs are generated, thus this option is currently the only one that works on Windows.

When running in regular mode `FileNotFoundError` raised:
```
...
  File "C:\Program Files\Python38\lib\subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\Python38\lib\subprocess.py", line 1307, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
  File "c:\Users\pasha\.vscode\extensions\ms-python.python-2020.8.108011\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydev_bundle\pydev_monkey.py", line 632, in new_CreateProcess
    return getattr(_subprocess, original_name)(app_name, cmd_line, *args)
FileNotFoundError: [WinError 2] The system cannot find the file specified
```

### PR DESCRIPTION
**Pull request by pashagolub**
Adds the possibility for `patronictl edit-config` custom pager support.

Fixes #1695 

Mentioned in [patroni-windows-packaging #16](https://github.com/cybertec-postgresql/patroni-windows-packaging/issues/16)

### CODE REVIEW

File: patroni/ctl.py [hunk 1]
```diff
@@ -1086,7 +1086,11 @@ class opts:
             side_by_side = False
             width = 80
             tab_width = 8
-        cdiff.markup_to_pager(cdiff.PatchStream(buf), opts)
+            wrap = True
+            pager = 'more.com' if (sys.platform == 'win32') else None
```
Comment (CyberDem0n) [inline patroni/ctl.py:1090]:
```suggestion
            if not find_executable('less'):
                pager = 'more.com' if sys.platform == 'win32' else 'more'
```

File: patroni/ctl.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 import datetime
 import dateutil.parser
 import dateutil.tz
-import cdiff
+import ydiff
```
Comment (CyberDem0n) [inline patroni/ctl.py:10]:
```suggestion
try:
    import ydiff
except ImportError:
    import cdiff as ydiff
```

And please move it after the line with `import yaml`

File: patroni/ctl.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-'''
+﻿'''
```
Comment (CyberDem0n) [inline patroni/ctl.py:1]: There is some unwanted character:
```
@@ -1,4 +1,4 @@
-'''
+<U+FEFF>'''
 Patroni Control
 '''
```
```suggestion
'''
```

File: requirements.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-urllib3[secure]>=1.19.1,!=1.21
+﻿urllib3[secure]>=1.19.1,!=1.21
```
Comment (CyberDem0n) [inline requirements.txt:1]: There is some unwanted character:
```
diff --git a/requirements.txt b/requirements.txt
index 5c30d3a..36cbae8 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,4 +1,4 @@
-urllib3[secure]>=1.19.1,!=1.21
+<U+FEFF>urllib3[secure]>=1.19.1,!=1.21
```

```suggestion
urllib3[secure]>=1.19.1,!=1.21
```

**Review by CyberDem0n — CHANGES_REQUESTED** [review]

File: patroni/ctl.py [hunk 5]
```diff
@@ -21,6 +20,10 @@
 import tempfile
 import time
 import yaml
+try:
+    import ydiff
+except ImportError:
+    import cdiff as ydiff
```
Comment (CyberDem0n) [inline patroni/ctl.py:26]:
```suggestion
try:
    from ydiff import markup_to_pager, PatchStream
except ImportError:  # pragma: no cover
    from cdiff import markup_to_pager, PatchStream
```

And please move it after the `from six.moves.urllib_parse import urlparse`

File: patroni/ctl.py [hunk 6]
```diff
@@ -1086,7 +1089,12 @@ class opts:
             side_by_side = False
             width = 80
             tab_width = 8
-        cdiff.markup_to_pager(cdiff.PatchStream(buf), opts)
+            wrap = True
+            if not find_executable('less'):
+                pager = 'more.com' if sys.platform == 'win32' else 'more'
+            pager_options = None
+
+        ydiff.markup_to_pager(ydiff.PatchStream(buf), opts)
```
Comment (CyberDem0n) [inline patroni/ctl.py:1097]:
```suggestion
            if find_executable('less'):
                pager = None
            else
                pager = 'more.com' if sys.platform == 'win32' else 'more'
            pager_options = None

        markup_to_pager(PatchStream(buf), opts)
```

**Review by CyberDem0n — CHANGES_REQUESTED** [review]
Unfortunately unit-tests are failing, therefore more changes are necessary (see comments).
And tests:
```diff
diff --git a/tests/test_ctl.py b/tests/test_ctl.py
index e63da17..d93c5d1 100644
--- a/tests/test_ctl.py
+++ b/tests/test_ctl.py
@@ -561,7 +561,7 @@ class TestCtl(unittest.TestCase):
         self.assertRaises(PatroniCtlException, apply_config_changes, before_editing, config, ['a'])

     @patch('sys.stdout.isatty', return_value=False)
-    @patch('cdiff.markup_to_pager')
+    @patch('patroni.ctl.markup_to_pager')
     def test_show_diff(self, mock_markup_to_pager, mock_isatty):
         show_diff("foo:\n  bar: 1\n", "foo:\n  bar: 2\n")
         mock_markup_to_pager.assert_not_called()
@@ -570,6 +570,9 @@ class TestCtl(unittest.TestCase):
         show_diff("foo:\n  bar: 1\n", "foo:\n  bar: 2\n")
         mock_markup_to_pager.assert_called_once()

+        with patch('patroni.ctl.find_executable', Mock(return_value=None)):
+            show_diff("foo:\n  bar: 1\n", "foo:\n  bar: 2\n")
+
         # Test that unicode handling doesn't fail with an exception
         show_diff(b"foo:\n  bar: \xc3\xb6\xc3\xb6\n".decode('utf-8'),
                   b"foo:\n  bar: \xc3\xbc\xc3\xbc\n".decode('utf-8'))
```

**Review by CyberDem0n — APPROVED** [review]

**Comment by CyberDem0n** [comment]
@mbanck, @df7cb this is sort of a "breaking" change, because there is no `python3-ydiff` package available.
A couple of years ago `cdiff` was renamed to `ydiff` and therefore no longer maintained.
In fact, interfaces didn't change much and patronictl will happily continue to work with `python3-cdiff`, so I believe it shouldn't become a big issue for you.

**Comment by sdudoladov** [comment]
:+1:

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**